### PR TITLE
ls: force fetching metadata when called with -L -Z

### PR DESCRIPTION
**Pull request by granquet**
The metadata are not used but it permits to use the same path as the long format to ensure we return 1 on invalid symlinks when ls is invoked with ls -L -Z

should solve https://github.com/uutils/coreutils/issues/2928

### CODE REVIEW

File: src/uu/ls/src/ls.rs [hunk 1]
```diff
@@ -2238,6 +2238,7 @@ fn display_items(items: &[PathData], config: &Config, out: &mut BufWriter<Stdout
             for item in items {
                 let context_len = item.security_context.len();
                 longest_context_len = context_len.max(longest_context_len);
+                item.md(out);
```
Comment (tertsdiepraam) [inline src/uu/ls/src/ls.rs:2241]: I think a comment explaining why this is here would be nice. I'm also thinking maybe it makes more sense to fix this in the `get_security_context` function?

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
I like the test and this the solution is  good. I just think it should be somewhere else, namely in the `get_security_context` function. That function triggers the error if selinux is enabled so it should probably also trigger the error if it doesn't.

File: src/uu/ls/src/ls.rs (hunk 1 above)
Comment (granquet) [reply src/uu/ls/src/ls.rs:2241]: I've looked at inserting something similar into `get_security_context()` but I really wanted to use the `md()` function from PathData as this is the "code path" that is taken in the case of `ls -L -l` and which generates the error.

At a first glance the patch looked a bit more complex to implement it inside `get_security_context()` as I don't have direct access to a `PathData()` or the `out` variable from `get_security_context()`.

I can try to make it happen with a bit of rework though :)

Comment (tertsdiepraam) [reply src/uu/ls/src/ls.rs:2241]: Just attempting to dereference in the `get_security_context` function should work right?

Comment (tertsdiepraam) [reply src/uu/ls/src/ls.rs:2241]: Essentially what I'm proposing is this:
- If security context is supported and the feature enabled: get the context and return the result
- If not and `must_dereference` is true:
    - perform a deref, if it fails, return that error
    - return Ok("?".into())
- Else just return Ok("?".into())

File: src/uu/ls/src/ls.rs [hunk 2]
```diff
@@ -3008,6 +3008,17 @@ fn display_inode(metadata: &Metadata) -> String {
 #[allow(unused_variables)]
 fn get_security_context(config: &Config, p_buf: &Path, must_dereference: bool) -> String {
     let substitute_string = "?".to_string();
+    if must_dereference {
```
Comment (tertsdiepraam) [inline src/uu/ls/src/ls.rs:3011]: A small comment explaining that this is for compatibility with GNU would be great

**Review by tertsdiepraam — COMMENTED** [review]
Code looks good! I just think it needs a comment to explain why we do this. Feel free to mention me when it's done and I'll take another look :)

File: src/uu/ls/src/ls.rs (hunk 2 above)
Comment (granquet) [reply src/uu/ls/src/ls.rs:3014]: @tertsdiepraam : I've added a small comment, awaiting your review :)

Comment (tertsdiepraam) [reply src/uu/ls/src/ls.rs:3014]: Nice! There's one crucial bit of info missing from that comment. You could add this to the end: "even when the system does not support SELinux"

**Comment by cakebaker** [comment]
Thanks for your PR :)

**Pull request merged**